### ISSUE TITLE
Persistent Sessions lead to a lot of ARJUNA012125 exceptions 

### ISSUE THREAD

**Issue by twobiers** [keycloak/keycloak#35602]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When using persistent offline sessions we see a couple of the following exceptions during regular operations. Eventually, it looks like the exceptions bubble up - I guess because of a retry backoff?
It is super hard to trace those errors properly due to the asynchronous writes.

<img width="657" alt="image" src="https://github.com/user-attachments/assets/d84a4378-029c-4d85-bf84-4fab6f5d7aff">

```
ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff0af4028f:8d21:67501ce9:278f95, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@3bf61a02 >
: jakarta.persistence.OptimisticLockException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update public.OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?
	at org.hibernate.internal.ExceptionConverterImpl.wrapStaleStateException(ExceptionConverterImpl.java:221)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:95)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:167)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:173)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1433)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:491)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2354)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1978)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:348)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:66)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:135)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1290)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
	at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:100)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.services.DefaultKeycloakSession.closeTransactionManager(DefaultKeycloakSession.java:396)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:361)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:400)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:273)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:263)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$3(PersistentSessionsWorker.java:89)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:102)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:93)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.process(PersistentSessionsWorker.java:86)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.run(PersistentSessionsWorker.java:69)
Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update public.OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?
	at org.hibernate.jdbc.Expectations.checkBatched(Expectations.java:80)
	at org.hibernate.jdbc.Expectation$RowCount.verifyOutcome(Expectation.java:174)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.checkRowCounts(BatchImpl.java:321)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:285)
	at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:67)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:264)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:242)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:188)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:674)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:511)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:414)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:41)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1429)
	... 26 more
```

Later on bubbling up a lot of ConstraintViolations.
```
ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff0af401a0:b123:67501d36:291ac3, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@2359920 >
: org.hibernate.exception.ConstraintViolationException: could not execute batch [Batch entry 0 insert into public.OFFLINE_USER_SESSION (BROKER_SESSION_ID,CREATED_ON,DATA,LAST_SESSION_REFRESH,REALM_ID,USER_ID,VERSION,OFFLINE_FLAG,USER_SESSION_ID) values ((NULL),('1733314746'::int4),('{"ipAddress":"REDACTED","authMethod":"openid-connect","rememberMe":false,"started":0,"notes":{"KC_DEVICE_NOTE":"REDACTED","AUTH_TIME":"1733314747","authenticators-completed":"{\"670834ea-01e9-4178-9588-87b8dd9e969b\":1733314746}"},"state":"LOGGED_IN"}'),('1733314746'::int4),('<REALM>'),('c37c9825-4b38-4ec5-b4b4-f24f1f564069'),('0'::int4),('0'),('4579f816-f1de-42ee-88d6-fefaceb79194')) was aborted: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
  Detail: Key (user_session_id, offline_flag)=(4579f816-f1de-42ee-88d6-fefaceb79194, 0) already exists.  Call getNextException to see other errors in the batch.] [insert into public.OFFLINE_USER_SESSION (BROKER_SESSION_ID,CREATED_ON,DATA,LAST_SESSION_REFRESH,REALM_ID,USER_ID,VERSION,OFFLINE_FLAG,USER_SESSION_ID) values (?,?,?,?,?,?,?,?,?)]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:97)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:58)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:294)
	at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:67)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:264)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:242)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:188)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:674)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:511)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:414)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:41)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1429)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:491)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2354)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1978)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:336)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:52)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:348)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:66)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:135)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1290)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:104)
	at io.quarkus.narayana.jta.runtime.NotifyingTransactionManager.commit(NotifyingTransactionManager.java:70)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:100)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.services.DefaultKeycloakSession.closeTransactionManager(DefaultKeycloakSession.java:396)
	at org.keycloak.services.DefaultKeycloakSession.close(DefaultKeycloakSession.java:361)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransactionWithResult(KeycloakModelUtils.java:400)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:273)
	at org.keycloak.models.utils.KeycloakModelUtils.runJobInTransaction(KeycloakModelUtils.java:263)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$2(PersistentSessionsWorker.java:98)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.lambda$process$3(PersistentSessionsWorker.java:96)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:102)
	at org.keycloak.common.util.Retry.executeWithBackoff(Retry.java:93)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.process(PersistentSessionsWorker.java:86)
	at org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker.run(PersistentSessionsWorker.java:69)
Caused by: java.sql.BatchUpdateException: Batch entry 0 insert into public.OFFLINE_USER_SESSION (BROKER_SESSION_ID,CREATED_ON,DATA,LAST_SESSION_REFRESH,REALM_ID,USER_ID,VERSION,OFFLINE_FLAG,USER_SESSION_ID) values ((NULL),('1733314746'::int4),('{"ipAddress":"REDACTED","authMethod":"openid-connect","rememberMe":false,"started":0,"notes":{"KC_DEVICE_NOTE":"REDACTED","","AUTH_TIME":"1733314747","authenticators-completed":"{\"670834ea-01e9-4178-9588-87b8dd9e969b\":1733314746}"},"state":"LOGGED_IN"}'),('1733314746'::int4),('<realm>'),('c37c9825-4b38-4ec5-b4b4-f24f1f564069'),('0'::int4),('0'),('4579f816-f1de-42ee-88d6-fefaceb79194')) was aborted: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
  Detail: Key (user_session_id, offline_flag)=(4579f816-f1de-42ee-88d6-fefaceb79194, 0) already exists.  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165)
	at org.postgresql.core.ResultHandlerDelegate.handleError(ResultHandlerDelegate.java:52)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2421)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:580)
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:888)
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:912)
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1739)
	at io.agroal.pool.wrapper.StatementWrapper.executeBatch(StatementWrapper.java:340)
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:279)
	... 38 more
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
  Detail: Key (user_session_id, offline_flag)=(4579f816-f1de-42ee-88d6-fefaceb79194, 0) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2733)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2420)
	... 44 more
```

### Version

26.0.6

### Regression

- [ ] The issue is a regression

### Expected behavior

I would not expect the exception to occur such frequently and it should allow tracing the session update causing the issue better

### Actual behavior

-

### How to Reproduce?

I guess running a large enough instance with concurrent requests would trigger it?

**Comment by andreb89** [comment]
We just updated from version 25.0.2 to 26.0.7 today and have the same Arjuna exceptions now (see below). With this update we have also starting using the new PERSISTENT_USER_SESSIONS feature. 

> {"timestamp":"2024-12-04T14:42:26.956053735+01:00","sequence":16209,"loggerClassName":"com.arjuna.ats.arjuna.logging.arjunaI18NLogger_$logger","loggerName":"com.arjuna.ats.arjuna","level":"WARN","message":"ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffff0a0000f6:a1c7:675024d4:15c22, org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization@1e38c27a >","threadName":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","threadId":60,"mdc":{},"ndc":"","hostName":"ecommerce-keycloak-1","processName":"/usr/lib/jvm/java-21-openjdk/bin/java","processId":1,"exception":{"refId":1,"exceptionType":"jakarta.persistence.OptimisticLockException","message":"Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?","frames":[{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"wrapStaleStateException","line":221},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":95},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":167},{"class":"org.hibernate.internal.ExceptionConverterImpl","method":"convert","line":173},{"class":"org.hibernate.internal.SessionImpl","method":"doFlush","line":1433},{"class":"org.hibernate.internal.SessionImpl","method":"managedFlush","line":491},{"class":"org.hibernate.internal.SessionImpl","method":"flushBeforeTransactionCompletion","line":2354},{"class":"org.hibernate.internal.SessionImpl","method":"beforeTransactionCompletion","line":1978},{"class":"org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl","method":"beforeTransactionCompletion","line":439},{"class":"org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl","method":"beforeCompletion","line":336},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl","method":"beforeCompletion","line":47},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization","method":"beforeCompletion","line":37},{"class":"com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple","method":"beforeCompletion","line":52},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"beforeCompletion","line":348},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"end","line":66},{"class":"com.arjuna.ats.arjuna.AtomicAction","method":"commit","line":135},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple","method":"commitAndDisassociate","line":1290},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction","method":"commit","line":104},{"class":"io.quarkus.narayana.jta.runtime.NotifyingTransactionManager","method":"commit","line":70},{"class":"org.keycloak.transaction.JtaTransactionWrapper","method":"commit","line":100},{"class":"org.keycloak.services.DefaultKeycloakTransactionManager","method":"commit","line":136},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"closeTransactionManager","line":396},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"close","line":361},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransactionWithResult","line":439},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransaction","line":312},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransaction","line":302},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"lambda$process$3","line":89},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":102},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":93},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"process","line":86},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"run","line":69}],"causedBy":{"exception":{"refId":2,"exceptionType":"org.hibernate.StaleStateException","message":"Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1; statement executed: update OFFLINE_USER_SESSION set BROKER_SESSION_ID=?,CREATED_ON=?,DATA=?,LAST_SESSION_REFRESH=?,REALM_ID=?,USER_ID=?,VERSION=? where OFFLINE_FLAG=? and USER_SESSION_ID=? and VERSION=?","frames":[{"class":"org.hibernate.jdbc.Expectations","method":"checkBatched","line":80},{"class":"org.hibernate.jdbc.Expectation$RowCount","method":"verifyOutcome","line":174},{"class":"org.hibernate.engine.jdbc.batch.internal.BatchImpl","method":"checkRowCounts","line":321},{"class":"org.hibernate.engine.jdbc.batch.internal.BatchImpl","method":"lambda$performExecution$2","line":285},{"class":"org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable","method":"forEachStatement","line":67},{"class":"org.hibernate.engine.jdbc.batch.internal.BatchImpl","method":"performExecution","line":264},{"class":"org.hibernate.engine.jdbc.batch.internal.BatchImpl","method":"execute","line":242},{"class":"org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl","method":"executeBatch","line":188},{"class":"org.hibernate.engine.spi.ActionQueue","method":"executeActions","line":674},{"class":"org.hibernate.engine.spi.ActionQueue","method":"executeActions","line":511},{"class":"org.hibernate.event.internal.AbstractFlushingEventListener","method":"performExecutions","line":414},{"class":"org.hibernate.event.internal.DefaultFlushEventListener","method":"onFlush","line":41},{"class":"org.hibernate.event.service.internal.EventListenerGroupImpl","method":"fireEventOnEachListener","line":127},{"class":"org.hibernate.internal.SessionImpl","method":"doFlush","line":1429},{"class":"org.hibernate.internal.SessionImpl","method":"managedFlush","line":491},{"class":"org.hibernate.internal.SessionImpl","method":"flushBeforeTransactionCompletion","line":2354},{"class":"org.hibernate.internal.SessionImpl","method":"beforeTransactionCompletion","line":1978},{"class":"org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl","method":"beforeTransactionCompletion","line":439},{"class":"org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl","method":"beforeCompletion","line":336},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl","method":"beforeCompletion","line":47},{"class":"org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization","method":"beforeCompletion","line":37},{"class":"com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple","method":"beforeCompletion","line":52},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"beforeCompletion","line":348},{"class":"com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator","method":"end","line":66},{"class":"com.arjuna.ats.arjuna.AtomicAction","method":"commit","line":135},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple","method":"commitAndDisassociate","line":1290},{"class":"com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction","method":"commit","line":104},{"class":"io.quarkus.narayana.jta.runtime.NotifyingTransactionManager","method":"commit","line":70},{"class":"org.keycloak.transaction.JtaTransactionWrapper","method":"commit","line":100},{"class":"org.keycloak.services.DefaultKeycloakTransactionManager","method":"commit","line":136},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"closeTransactionManager","line":396},{"class":"org.keycloak.services.DefaultKeycloakSession","method":"close","line":361},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransactionWithResult","line":439},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransaction","line":312},{"class":"org.keycloak.models.utils.KeycloakModelUtils","method":"runJobInTransaction","line":302},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"lambda$process$3","line":89},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":102},{"class":"org.keycloak.common.util.Retry","method":"executeWithBackoff","line":93},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"process","line":86},{"class":"org.keycloak.models.sessions.infinispan.changes.PersistentSessionsWorker$BatchWorker","method":"run","line":69}]}}}}

**Comment by twobiers** [comment]
> I'm pretty sure that wasn't the case in 25.0.6 (which didn't persist user sessions by default, right?)

I think if you have regular OIDC offline sessions, they will be persisted using the same underlying mechanism.

**Comment by famod** [comment]
> > I'm pretty sure that wasn't the case in 25.0.6 (which didn't persist user sessions by default, right?)
> 
> I think if you have regular OIDC offline sessions, they will be persisted using the same underlying mechanism.

Yeah, right - but I'm not using the offline_access scope.

**Comment by famod** [comment]
Unsurprisingly, setting `--features-disabled=persistent-user-sessions` prevents this issue in my setup.

**Comment by ahus1** [comment]
Closing as a duplicate of #35047

**Comment by famod** [comment]
@ahus1 sorry about commenting a closed issue, but it seems #35047 / #36310 has not fixed `duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"` as I'm stills seeing it with Keycloak 26.1.3.

**Comment by antonmartsenyuk** [comment]
### 📨 **Subject:** Duplicate Key Violation When Reauthenticating After Account Deletion via Google  

**Issue Description:**  
We encountered an issue in Keycloak **26.1.4** where users cannot immediately reauthenticate after deleting their account when using Google (or another external identity provider). The error message indicates a **duplicate key violation** in the database.  

### 🔄 **Steps to Reproduce:**  
1. Configure authentication via **Google** (or another external provider).  
2. Navigate to **Authentication → Required Actions** and enable **"Delete Account"**.  
3. Go to **Realm Settings → User Registration → Default Roles**, add the **"Delete Account"** role.  
4. Log in with **Google**, then delete the account through the UI.  
5. Immediately attempt to log in again with Google.  
6. The following error occurs:  

```
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"
Detail: Key (user_session_id, offline_flag)=(4eb36d91-6d49-442f-8fad-9be63896d89b, 0) already exists.
```

### 🔍 **Expected Behavior:**  
The user should be able to reauthenticate successfully as if they are logging in for the first time.  

### 🛠 **Additional Information:**  
- **Keycloak Version:** 26.1.4  
- **Database:** PostgreSQL  
- **Deployment:** Running in a container based on `bitnami/keycloak:26.1.4`  

### ⚠ **Important Note:**  
The issue **resolves itself after approximately 5 minutes**, allowing the user to log in successfully again. However, the immediate failure creates a poor user experience and may indicate an underlying issue with session or user data cleanup.  

### ❓ **Possible Cause:**  
It seems that after account deletion, some session-related data (e.g., `user_session_id`) is not fully removed from the database. When the user tries to log in again immediately, Keycloak attempts to reuse an existing session, resulting in a **constraint violation** in `offline_user_session`.  

We would appreciate any insights or a potential fix for this issue! 🙏

**Comment by antonmartsenyuk** [comment]
https://github.com/keycloak/keycloak/issues/38671

**Comment by Jayter** [comment]
@ahus1 
Do you mind revisiting this issue?
The problem itself is `duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"`, not a backoff time that was fixed in https://github.com/keycloak/keycloak/issues/35047.
We started facing this issue in multiple circumstances after Keycloak upgrade from 24.0.5 to 26.1.2. It happens on the 2-nd step of Authorization Code Flow. Situations are different from case to case: some of situations are concurrent logins, in some cases customer can login for the 2-nd/3rd time during 1 hour (session lifetime), in some cases it is the first login during last 2 days and it also fails.
We are about to disable this feature for now, but I have searched through Keycloak issues and I have not found any signs that this issue is on track and about to be fixed.

**Comment by Jayter** [comment]
> [@ahus1](https://github.com/ahus1) Do you mind revisiting this issue? The problem itself is `duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"`, not a backoff time that was fixed in [#35047](https://github.com/keycloak/keycloak/issues/35047). We started facing this issue in multiple circumstances after Keycloak upgrade from 24.0.5 to 26.1.2. It happens on the 2-nd step of Authorization Code Flow. Situations are different from case to case: some of situations are concurrent logins, in some cases customer can login for the 2-nd/3rd time during 1 hour (session lifetime), in some cases it is the first login during last 2 days and it also fails. We are about to disable this feature for now, but I have searched through Keycloak issues and I have not found any signs that this issue is on track and about to be fixed.

**Comment by twobiers** [comment]
That issue seems to be addressed now in https://github.com/keycloak/keycloak/pull/39501/

**Comment by ahus1** [comment]
@twobiers - issue #39501 addresses that those error that Keycloak successfully recovers from by retrying should now longer appear in the logs. That is part of 26.2.4 which should be released soon. 

If you still see errors in the logs after that, please raise new issues.

**Comment by Sieboldianus** [comment]
I am on Keycloak 26.2.5 and still reliable reproduce `duplicate key value violates unique constraint "constraint_offl_us_ses_pk2"` error. Once I added `RUN /opt/keycloak/bin/kc.sh build --features-disabled=persistent-user-sessions` to my Dockerfile, the issue disappeared.

**Comment by ahus1** [comment]
@Sieboldianus - those exception will still happen in 26.2.5 when someone double-submits a login form. Some fixed were added in #39759 which will be part of 26.2.6

Disabling persistent user sessions just hides the symptoms. 

We are also investigating #40374 for a setup with brokering, possibly when the request from the external IDP arrive multiple times (double submit on side of the IDP). 

If you think you have an issue related to #40374, please comment there. If you think there are issues beyond #39759, please raise a new issue.
